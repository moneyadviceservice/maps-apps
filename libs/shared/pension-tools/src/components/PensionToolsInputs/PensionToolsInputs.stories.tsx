import { StoryFn } from '@storybook/react';

import {
  mockerrors,
  mockQuestions,
} from '../PensionPotCalculator/mockTestData';
import {
  PensionToolsInputs,
  PensionToolsInputsProps,
} from './PensionToolsInputs';

const StoryProps = {
  title: 'Components/PENSION-TOOLS/PensionToolsInputs',
  component: PensionToolsInputs,
};

const Template: StoryFn<PensionToolsInputsProps> = (args) => (
  <PensionToolsInputs {...args} />
);

export const Default = Template.bind({});
Default.args = {
  field: mockQuestions[0],
  errors: mockerrors,
  queryData: {
    pot: '1000',
    age: '30',
  },
};

export default StoryProps;
