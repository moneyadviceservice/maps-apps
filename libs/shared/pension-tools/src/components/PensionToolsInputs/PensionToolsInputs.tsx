import { ChangeEvent, ReactNode } from 'react';

import { twMerge } from 'tailwind-merge';

import { Errors } from '@maps-react/common/components/Errors';
import { MoneyInput } from '@maps-react/form/components/MoneyInput';
import { NumberInput } from '@maps-react/form/components/NumberInput';
import { ErrorObject, Question } from '@maps-react/form/types';
import useTranslation from '@maps-react/hooks/useTranslation';
import { DataFromQuery } from '@maps-react/utils/pageFilter';

import { addUnitToAriaLabel } from '../../utils/addUnitToAriaLabel';
import { isInputAllowedDefault } from '../../utils/inputValidation';

export type PensionToolsInputsProps = {
  field: Question;
  errors: ErrorObject;
  queryData: DataFromQuery;
  isAllowed: (value: { floatValue: number | undefined }) => boolean;
  value: string;
  onChange?: (e: ChangeEvent<HTMLInputElement>, field: Question) => void;
};

export const PensionToolsInputs = ({
  field,
  errors,
  queryData,
  onChange,
  isAllowed,
  value,
}: PensionToolsInputsProps) => {
  const { z } = useTranslation();

  const getErrorMessage = (field: Question, errors: ErrorObject) => {
    const HTMLvariant = (field.errors as Record<string, ReactNode>)[
      `${errors[field.type].type}HTML`
    ];
    if (HTMLvariant) {
      return HTMLvariant;
    }
    return (field.errors as Record<string, string>)[errors[field.type].type];
  };

  const hasError =
    field?.errors &&
    errors[field?.type] &&
    (field?.errors as Record<string, string>)[errors[field?.type]?.type];

  const hasRequired =
    field?.errors &&
    (field?.errors as Record<string, string>)['required'] !== undefined;

  return (
    <Errors
      errors={hasError ? ['error'] : []}
      className={twMerge(hasError ? 'pl-4' : '')}
    >
      <fieldset className="mb-4">
        <label
          htmlFor={field.type}
          className={twMerge(
            'text-2xl font-medium text-gray-800 inline-flex mb-2',
            !!field.description && 'mb-0',
          )}
        >
          {field.title}
        </label>
        {field.description}
        {hasError && field?.errors && (
          <div className="mb-2 text-red-700" aria-describedby={field.type}>
            {getErrorMessage(field, errors)}
          </div>
        )}
        {field.type === 'age' && (
          <div className="flex">
            <NumberInput
              data-testid={field.type}
              aria-required={hasRequired || undefined}
              id={field.type}
              key={field.type}
              name={field.type}
              className="border-gray-600 border-r-0 h-[49px] ml-0 max-w-[307px]"
              value={queryData[field.type]}
              onChange={(e) => {
                onChange && onChange(e, field);
              }}
              aria-label={addUnitToAriaLabel(field.title, 'years', z)}
            />
            <span
              aria-hidden
              className="bg-gray-100 py-2 px-3.5 rounded-r border-gray-400 border border-solid min-w-[92px] mr-2 flex items-center"
            >
              {z({
                en: 'Years',
                cy: 'flynedd',
              })}
            </span>
          </div>
        )}
        {field.type !== 'age' && (
          <MoneyInput
            data-testid={field.type}
            aria-required={hasRequired || undefined}
            id={field.type}
            key={field.type}
            name={field.type}
            className="mr-2 border-gray-600 border-l-0 max-w-[360px]"
            isAllowed={({ floatValue }) =>
              isAllowed({ floatValue }) && isInputAllowedDefault({ floatValue })
            }
            value={value}
            onChange={(e: ChangeEvent<HTMLInputElement>) => {
              onChange && onChange(e, field);
            }}
            aria-label={addUnitToAriaLabel(field.title, 'pounds', z)}
          />
        )}
      </fieldset>
    </Errors>
  );
};
