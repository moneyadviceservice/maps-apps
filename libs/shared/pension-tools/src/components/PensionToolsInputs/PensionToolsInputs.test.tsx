import { fireEvent, render } from '@testing-library/react';

import {
  mockerrors,
  mockQuestions,
} from '../PensionPotCalculator/mockTestData';
import { PensionToolsInputs } from './PensionToolsInputs';

jest.mock('next/router', () => ({
  useRouter: jest.fn().mockReturnValue({
    push: jest.fn(),
    query: {
      language: 'en',
    },
    events: {
      on: jest.fn(),
      off: jest.fn(),
    },
  }),
}));

describe('PensionToolsInputs', () => {
  it('should render an input fields with label, description, error message and values successfully ', () => {
    mockQuestions.forEach((question) => {
      const { container } = render(
        <PensionToolsInputs
          field={question}
          errors={mockerrors}
          queryData={{
            pot: '1000',
            age: '30',
          }}
          isAllowed={function ({ floatValue }): boolean {
            return !!floatValue && floatValue > 0;
          }}
          value={''}
          onChange={jest.fn()}
        />,
      );
      expect(container).toMatchSnapshot();
    });
  });

  it('should render an input pot fields successfully without any value and enter value', () => {
    mockQuestions.forEach((question) => {
      const { container } = render(
        <PensionToolsInputs
          field={question}
          errors={mockerrors}
          queryData={{
            pot: '1000',
            age: '30',
          }}
          isAllowed={function ({ floatValue }): boolean {
            return !!floatValue && floatValue > 0;
          }}
          value={''}
          onChange={jest.fn()}
        />,
      );

      fireEvent.change(container.querySelector('input') as Element, {
        target: { value: '1000' },
      });

      expect(container).toMatchSnapshot();
    });
  });
});
