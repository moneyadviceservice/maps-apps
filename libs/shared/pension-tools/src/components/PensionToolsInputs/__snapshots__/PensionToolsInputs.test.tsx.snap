// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PensionToolsInputs should render an input fields with label, description, error message and values successfully  1`] = `
<div>
  <div
    class="border-0 border-l-4 border-red-700 border-solid pl-4"
    data-testid="errors"
  >
    <fieldset
      class="mb-4"
    >
      <label
        class="text-2xl font-medium text-gray-800 inline-flex mb-0"
        for="pot"
      >
        How much is in your pot?
      </label>
      <p
        class="text-base text-gray-400 mb-2"
      >
        Le Lorem Ipsum est simplement du faux texte employé dans la composition et la mise en page avant impression.
      </p>
      <div
        aria-describedby="pot"
        class="mb-2 text-red-700"
      >
        Enter a figure
      </div>
      <div
        class="flex"
      >
        <span
          aria-hidden="true"
          class="bg-gray-100 py-2 px-3 rounded-l border-gray-400 border border-solid leading-[31px] text-[18px]"
        >
          £
        </span>
        <input
          aria-label="How much is in your pot? In pounds."
          aria-required="true"
          class="tool-field w-full border pl-2 ml-0 h-auto focus:outline-purple-700 focus:shadow-focus-outline mr-2 border-gray-600 border-l-0 max-w-[360px]"
          data-testid="pot"
          id="pot"
          inputmode="numeric"
          name="pot"
          type="text"
          value=""
        />
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`PensionToolsInputs should render an input fields with label, description, error message and values successfully  2`] = `
<div>
  <div
    class="border-0 border-l-4 border-red-700 border-solid pl-4"
    data-testid="errors"
  >
    <fieldset
      class="mb-4"
    >
      <label
        class="text-2xl font-medium text-gray-800 inline-flex mb-0"
        for="age"
      >
        How old are you?
      </label>
      <p
        class="text-base text-gray-400 mb-2"
      >
        Age description
      </p>
      <div
        aria-describedby="age"
        class="mb-2 text-red-700"
      >
        Enter a figure
      </div>
      <div
        class="flex"
      >
        <input
          aria-label="How old are you? In years."
          aria-required="true"
          class="tool-field w-full border pl-2 focus:outline-purple-700 focus:shadow-focus-outline border-gray-600 border-r-0 h-[49px] ml-0 max-w-[307px]"
          data-testid="age"
          id="age"
          inputmode="numeric"
          name="age"
          type="text"
          value="30"
        />
        <span
          aria-hidden="true"
          class="bg-gray-100 py-2 px-3.5 rounded-r border-gray-400 border border-solid min-w-[92px] mr-2 flex items-center"
        >
          Years
        </span>
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`PensionToolsInputs should render an input pot fields successfully without any value and enter value 1`] = `
<div>
  <div
    class="border-0 border-l-4 border-red-700 border-solid pl-4"
    data-testid="errors"
  >
    <fieldset
      class="mb-4"
    >
      <label
        class="text-2xl font-medium text-gray-800 inline-flex mb-0"
        for="pot"
      >
        How much is in your pot?
      </label>
      <p
        class="text-base text-gray-400 mb-2"
      >
        Le Lorem Ipsum est simplement du faux texte employé dans la composition et la mise en page avant impression.
      </p>
      <div
        aria-describedby="pot"
        class="mb-2 text-red-700"
      >
        Enter a figure
      </div>
      <div
        class="flex"
      >
        <span
          aria-hidden="true"
          class="bg-gray-100 py-2 px-3 rounded-l border-gray-400 border border-solid leading-[31px] text-[18px]"
        >
          £
        </span>
        <input
          aria-label="How much is in your pot? In pounds."
          aria-required="true"
          class="tool-field w-full border pl-2 ml-0 h-auto focus:outline-purple-700 focus:shadow-focus-outline mr-2 border-gray-600 border-l-0 max-w-[360px]"
          data-testid="pot"
          id="pot"
          inputmode="numeric"
          name="pot"
          type="text"
          value="1,000"
        />
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`PensionToolsInputs should render an input pot fields successfully without any value and enter value 2`] = `
<div>
  <div
    class="border-0 border-l-4 border-red-700 border-solid pl-4"
    data-testid="errors"
  >
    <fieldset
      class="mb-4"
    >
      <label
        class="text-2xl font-medium text-gray-800 inline-flex mb-0"
        for="age"
      >
        How old are you?
      </label>
      <p
        class="text-base text-gray-400 mb-2"
      >
        Age description
      </p>
      <div
        aria-describedby="age"
        class="mb-2 text-red-700"
      >
        Enter a figure
      </div>
      <div
        class="flex"
      >
        <input
          aria-label="How old are you? In years."
          aria-required="true"
          class="tool-field w-full border pl-2 focus:outline-purple-700 focus:shadow-focus-outline border-gray-600 border-r-0 h-[49px] ml-0 max-w-[307px]"
          data-testid="age"
          id="age"
          inputmode="numeric"
          name="age"
          type="text"
          value="1000"
        />
        <span
          aria-hidden="true"
          class="bg-gray-100 py-2 px-3.5 rounded-r border-gray-400 border border-solid min-w-[92px] mr-2 flex items-center"
        >
          Years
        </span>
      </div>
    </fieldset>
  </div>
</div>
`;
